File: sop-java/src/main/java/sop/Verification.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop;

import sop.enums.SignatureMode;
import sop.util.Optional;
import sop.util.UTCUtil;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.Date;

/**
 * Class bundling information about a verified signature.
 */
public class Verification {

    private final Date creationTime;
    private final String signingKeyFingerprint;
    private final String signingCertFingerprint;
    private final Optional<SignatureMode> signatureMode;
    private final Optional<String> description;

    private static final String MODE = "mode:";

    /**
     * Create a new {@link Verification} without mode and description.
     *
     * @param creationTime signature creation time
     * @param signingKeyFingerprint fingerprint of the signing (sub-) key
     * @param signingCertFingerprint fingerprint of the certificate
     */
    public Verification(@Nonnull Date creationTime,
                        @Nonnull String signingKeyFingerprint,
                        @Nonnull String signingCertFingerprint) {
        this(creationTime, signingKeyFingerprint, signingCertFingerprint, Optional.ofEmpty(), Optional.ofEmpty());
    }

    /**
     * Create a new {@link Verification}.
     *
     * @param creationTime signature creation time
     * @param signingKeyFingerprint fingerprint of the signing (sub-) key
     * @param signingCertFingerprint fingerprint of the certificate
     * @param signatureMode signature mode (optional, may be <pre>null</pre>)
     * @param description free-form description, e.g. <pre>certificate from dkg.asc</pre> (optional, may be <pre>null</pre>)
     */
    public Verification(@Nonnull Date creationTime,
                        @Nonnull String signingKeyFingerprint,
                        @Nonnull String signingCertFingerprint,
                        @Nullable SignatureMode signatureMode,
                        @Nullable String description) {
        this(
                creationTime,
                signingKeyFingerprint,
                signingCertFingerprint,
                Optional.ofNullable(signatureMode),
                Optional.ofNullable(nullSafeTrim(description))
        );
    }

    private Verification(@Nonnull Date creationTime,
                         @Nonnull String signingKeyFingerprint,
                         @Nonnull String signingCertFingerprint,
                         @Nonnull Optional<SignatureMode> signatureMode,
                         @Nonnull Optional<String> description) {
        this.creationTime = creationTime;
        this.signingKeyFingerprint = signingKeyFingerprint;
        this.signingCertFingerprint = signingCertFingerprint;
        this.signatureMode = signatureMode;
        this.description = description;
    }

    private static String nullSafeTrim(@Nullable String string) {
        if (string == null) {
            return null;
        }
        return string.trim();
    }

    @Nonnull
    public static Verification fromString(@Nonnull String toString) {
        String[] split = toString.trim().split(" ");
        if (split.length < 3) {
            throw new IllegalArgumentException("Verification must be of the format 'UTC-DATE OpenPGPFingerprint OpenPGPFingerprint [mode] [info]'");
        }

        if (split.length == 3) {
            return new Verification(
                    UTCUtil.parseUTCDate(split[0]), // timestamp
                    split[1],                       // key FP
                    split[2]                        // cert FP
            );
        }

        SignatureMode mode = null;
        int index = 3;
        if (split[index].startsWith(MODE)) {
            mode = SignatureMode.valueOf(split[3].substring(MODE.length()));
            index++;
        }

        StringBuilder sb = new StringBuilder();
        for (int i = index; i < split.length; i++) {
            if (sb.length() != 0) {
                sb.append(' ');
            }
            sb.append(split[i]);
        }

        return new Verification(
                UTCUtil.parseUTCDate(split[0]),         // timestamp
                split[1],                               // key FP
                split[2],                               // cert FP
                mode,                                   // signature mode
                sb.length() != 0 ? sb.toString() : null // description
        );
    }

    /**
     * Return the signatures' creation time.
     *
     * @return signature creation time
     */
    @Nonnull
    public Date getCreationTime() {
        return creationTime;
    }

    /**
     * Return the fingerprint of the signing (sub)key.
     *
     * @return signing key fingerprint
     */
    @Nonnull
    public String getSigningKeyFingerprint() {
        return signingKeyFingerprint;
    }

    /**
     * Return the fingerprint fo the signing certificate.
     *
     * @return signing certificate fingerprint
     */
    @Nonnull
    public String getSigningCertFingerprint() {
        return signingCertFingerprint;
    }

    /**
     * Return the mode of the signature.
     * Optional, may return <pre>null</pre>.
     *
     * @return signature mode
     */
    @Nonnull
    public Optional<SignatureMode> getSignatureMode() {
        return signatureMode;
    }

    /**
     * Return an optional description.
     * Optional, may return <pre>null</pre>.
     *
     * @return description
     */
    @Nonnull
    public Optional<String> getDescription() {
        return description;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder()
                .append(UTCUtil.formatUTCDate(getCreationTime()))
                .append(' ')
                .append(getSigningKeyFingerprint())
                .append(' ')
                .append(getSigningCertFingerprint());

        if (signatureMode.isPresent()) {
            sb.append(' ').append(MODE).append(signatureMode.get());
        }

        if (description.isPresent()) {
            sb.append(' ').append(description.get());
        }

        return sb.toString();
    }

    @Override
    public int hashCode() {
        return toString().hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof Verification)) {
            return false;
        }
        Verification other = (Verification) obj;
        return toString().equals(other.toString());
    }
}

File: sop-java/src/main/kotlin/sop/Verification.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop

import java.text.ParseException
import java.util.Date
import sop.enums.SignatureMode
import sop.util.Optional
import sop.util.UTCUtil

/**
 * Metadata about a verified signature.
 *
 * @param creationTime creation time of the signature
 * @param signingKeyFingerprint fingerprint of the (sub-)key that issued the signature
 * @param signingCertFingerprint fingerprint of the certificate that contains the signing key
 * @param signatureMode optional signature mode (text/binary)
 * @param jsonOrDescription arbitrary text or JSON data
 */
data class Verification(
    val creationTime: Date,
    val signingKeyFingerprint: String,
    val signingCertFingerprint: String,
    val signatureMode: Optional<SignatureMode>,
    val jsonOrDescription: Optional<String>
) {

    @JvmOverloads
    constructor(
        creationTime: Date,
        signingKeyFingerprint: String,
        signingCertFingerprint: String,
        signatureMode: SignatureMode? = null,
        description: String? = null
    ) : this(
        creationTime,
        signingKeyFingerprint,
        signingCertFingerprint,
        Optional.ofNullable(signatureMode),
        Optional.ofNullable(description?.trim()))

    @JvmOverloads
    constructor(
        creationTime: Date,
        signingKeyFingerprint: String,
        signingCertFingerprint: String,
        signatureMode: SignatureMode? = null,
        json: JSON,
        jsonSerializer: JSONSerializer
    ) : this(
        creationTime,
        signingKeyFingerprint,
        signingCertFingerprint,
        Optional.ofNullable(signatureMode),
        Optional.of(jsonSerializer.serialize(json)))

    @Deprecated("Replaced by jsonOrDescription", replaceWith = ReplaceWith("jsonOrDescription"))
    val description = jsonOrDescription

    /** This value is `true` if the [Verification] contains extension JSON. */
    val containsJson: Boolean =
        jsonOrDescription.get()?.trim()?.let { it.startsWith("{") && it.endsWith("}") } ?: false

    /**
     * Attempt to parse the [jsonOrDescription] field using the provided [JSONParser] and return the
     * result. This method returns `null` if parsing fails.
     *
     * @param parser [JSONParser] implementation
     * @return successfully parsed [JSON] POJO or `null`.
     */
    fun getJson(parser: JSONParser): JSON? {
        return jsonOrDescription.get()?.let {
            try {
                parser.parse(it)
            } catch (e: ParseException) {
                null
            }
        }
    }

    override fun toString(): String =
        "${UTCUtil.formatUTCDate(creationTime)} $signingKeyFingerprint $signingCertFingerprint" +
            (if (signatureMode.isPresent) " mode:${signatureMode.get()}" else "") +
            (if (jsonOrDescription.isPresent) " ${jsonOrDescription.get()}" else "")

    companion object {
        @JvmStatic
        fun fromString(string: String): Verification {
            val split = string.trim().split(" ")
            require(split.size >= 3) {
                "Verification must be of the format 'UTC-DATE OpenPGPFingerprint OpenPGPFingerprint [mode] [info]'."
            }
            if (split.size == 3) {
                return Verification(parseUTCDate(split[0]), split[1], split[2])
            }

            var index = 3
            val mode =
                if (split[3].startsWith("mode:")) {
                    index += 1
                    SignatureMode.valueOf(split[3].substring("mode:".length))
                } else null

            val description = split.subList(index, split.size).joinToString(" ").ifBlank { null }

            return Verification(
                parseUTCDate(split[0]),
                split[1],
                split[2],
                Optional.ofNullable(mode),
                Optional.ofNullable(description))
        }

        @JvmStatic
        private fun parseUTCDate(string: String): Date {
            return try {
                UTCUtil.parseUTCDate(string)
            } catch (e: ParseException) {
                throw IllegalArgumentException("Malformed UTC timestamp.", e)
            }
        }
    }

    /**
     * POJO data class representing JSON metadata.
     *
     * @param signers list of supplied CERTS objects that could have issued the signature,
     *   identified by the name given on the command line.
     * @param comment a freeform UTF-8 encoded text describing the verification
     * @param ext an extension object containing arbitrary, implementation-specific data
     */
    data class JSON(val signers: List<String>, val comment: String?, val ext: Any?) {

        /** Create a JSON object with only a list of signers. */
        constructor(signers: List<String>) : this(signers, null, null)

        /** Create a JSON object with only a single signer. */
        constructor(signer: String) : this(listOf(signer))
    }

    /** Interface abstracting a JSON parser that parses [JSON] POJOs from single-line strings. */
    fun interface JSONParser {
        /**
         * Parse a [JSON] POJO from the given single-line [string]. If the string does not represent
         * a JSON object matching the [JSON] definition, this method throws a [ParseException].
         *
         * @param string [String] representation of the [JSON] object.
         * @return parsed [JSON] POJO
         * @throws ParseException if the [string] is not a JSON string representing the [JSON]
         *   object.
         */
        @Throws(ParseException::class) fun parse(string: String): JSON
    }

    /**
     * Interface abstracting a JSON serializer that converts [JSON] POJOs into single-line JSON
     * strings.
     */
    fun interface JSONSerializer {

        /**
         * Serialize the given [JSON] object into a single-line JSON string.
         *
         * @param json JSON POJO
         * @return JSON string
         */
        fun serialize(json: JSON): String
    }
}
